File: Laboratory_7/Laboratory_2/Kotlin/task2.kt
import java.util.Scanner

fun printBottles(bottle: String, count: Int) {
    /* Выводит бутылки на экран. */
    for (i in 0 until count) {
        if (i % 3 == 0) println()
        print("$bottle ")
    }
}

fun drinkSimulation(n: Int, m: Int): Pair<Int, Int> {
    /* Показывает все шаги питья. */
    var full = n
    var empty = 0
    var totalDrunk = 0
    var iterations = 0

    val fullBottle = "●"
    val emptyBottle = "○"

    // Выведем дано
    println("Дано")
    printBottles(fullBottle, full)
    println("\n")

    while (true) {
        // Шаг 1: выпить все полные бутылки
        if (full > 0) {
            totalDrunk += full
            empty += full
            full = 0
            iterations++

            // Выведем все выпитые бутылки
            print("$iterations шаг")
            printBottles(emptyBottle, empty)
            println("\n")

            if (empty < m) break // условие завершения

            // Шаг 2: обмен пустых бутылок на полные
            full = empty / m
            empty %= m
            iterations++

            // Выведем бутылки после обмена
            print("$iterations шаг")
            printBottles(fullBottle, full)
            printBottles(emptyBottle, empty)
            println("\n")
        }
    }

    return Pair(totalDrunk, iterations)
}

fun main() {
    val scanner = Scanner(System.`in`)
    print("Введите количество полных бутылок и количество бутылок, которых достаточно для обмена: ")
    val n = scanner.nextInt()
    val m = scanner.nextInt()

    println("Input: $n $m")

    val (total, iters) = drinkSimulation(n, m)

    println("Output (total, iters): $total $iters")
}

File: LaboratoryWork/Semester_2/Laboratory_7/Laboratory_2/C#/task2/task2.cs
using System;

class Task2 {
    static void PrintBottles(string bottle, int count) {
        /* Выводит бутылки на экран. */
        for (int i = 0; i < count; i++) {
            if (i % 3 == 0) Console.WriteLine();
            Console.Write(bottle + " ");
        }
    }

    static (int, int) DrinkSimulation(int n, int m) {
        /* Показывает все шаги питья. */
        int full = n;
        int empty = 0;
        int totalDrunk = 0;
        int iterations = 0;

        string fullBottle = "●";
        string emptyBottle = "○";

        // Выведем дано
        Console.Write("Дано");
        PrintBottles(fullBottle, full);
        Console.WriteLine("\n");

        while (true) {
            // Шаг 1: выпить все полные бутылки
            if (full > 0) {
                totalDrunk += full;
                empty += full;
                full = 0;
                iterations++;

                // Выведем все выпитые бутылки
                Console.Write(iterations + " шаг");
                PrintBottles(emptyBottle, empty);
                Console.WriteLine("\n");

                if (empty < m) break;  // условие завершения

                // Шаг 2: обмен пустых бутылок на полные
                full = empty / m;
                empty = empty % m;
                iterations++;

                // Выведем бутылки после обмена 
                Console.Write(iterations + " шаг");
                PrintBottles(fullBottle, full);
                PrintBottles(emptyBottle, empty);
                Console.WriteLine("\n");
            }
        }

        return (totalDrunk, iterations);
    }

    static void Main() {
        Console.Write("Введите количество полных бутылок и количество бутылок, которых достаточно для обмена: ");
        string[] input = Console.ReadLine().Split();

        if (input.Length < 2 || !int.TryParse(input[0], out int n) || !int.TryParse(input[1], out int m)) {
            Console.WriteLine("Ошибка ввода! Введите два целых числа.");
            return;
        }

        Console.WriteLine("Input: " + n + " " + m);

        (int total, int iters) = DrinkSimulation(n, m);

        Console.WriteLine("Output (total, iters): " + total + " " + iters);
    }
}